FerramAerospaceResearch/FARAeroComponents/VehicleOcclusion.cs

```diff
@@ -0,0 +1,828 @@
+                                    (Vector3)(v.transform.worldToLocalMatrix * (Vector3)v.velocityD);
+            Vector3 localSunVec = v.transform.worldToLocalMatrix * (Planetarium.fetch.Sun.transform.position - v.transform.position);
+            Vector3 bodyVec = v.transform.worldToLocalMatrix * (Vector3)(Vessel.mainBody.position - v.transform.position);
+            dirs.Add(localVelocity.normalized);
+            dirs.Add(localSunVec.normalized);
+            dirs.Add(bodyVec.normalized);
+        }
+
+        public void FixedUpdateEarly()
+        {
+            if (!OnValidPhysicsVessel)
+                return;
+
+            // We must call this to pre-calculate the nearest quaternions for sun and convection.
+            // It may return an empty work queue if we are not otherwise running, which is fine.
+            if (state == State.Initialized)
+                state = State.Running;
+            if (state == State.Running || state == State.Completed)
+            {
+                SetupDefaultDirs(directionList, Vessel);
+                LaunchJobs(Vessel, directionPreprocessInfos, directionList, occlusionPointsList, 0);
+                if (!PassStarted)
+                {
+                    PassStarted = true;
+                    JobsInCurrentPass = 0;
+                }
+                JobsInCurrentPass += jobsInProgress;
+            }
+        }
+
+        public void FixedUpdate()
+        {
+            PassStarted = false;
+            if (!OnValidPhysicsVessel)
+                return;
+            if (state == State.Running)
+                HandleJobCompletion(raycastJobTracker, FixedUpdateMetric);
+        }
+
+        public void UpdateEarly()
+        {
+            if (!OnValidPhysicsVessel)
+                return;
+            if (state == State.Running)
+            {
+                int threadCount = System.Environment.ProcessorCount - 1;
+                int availableJobs = math.max(OcclusionSettings.MaxJobs - JobsInCurrentPass, 0);
+                int updateJobs = math.min(availableJobs, math.max(1, Settings.OcclusionSettings.JobsPerThread * threadCount));
+                SetupDefaultDirs(directionList, Vessel);
+                LaunchJobs(Vessel, directionPreprocessInfos, directionList, occlusionPointsList, updateJobs);
+                if (!PassStarted)
+                {
+                    PassStarted = true;
+                    JobsInCurrentPass = 0;
+                }
+                JobsInCurrentPass += jobsInProgress;
+            }
+        }
+
+        public void LateUpdateComplete()
+        {
+            PassStarted = false;
+            if (!OnValidPhysicsVessel)
+                return;
+            if (state == State.Running && jobsInProgress == 0 && processedQuaternionsMap.Length == Quaternions.Length)
+                state = State.Completed;
+            else if (state == State.Running)
+                HandleJobCompletion(raycastJobTracker, UpdateMetric);
```

Comment:
Does this need to run in `LateUpdate` when the calculations only affect physics?

Reply:
Yes.  The goal is to actually do most of the compute to fill in the cache during the Update cycle, and only do things in FixedUpdate that will be required immediately and weren't already computed.  So I'm trying to allow as much time as possible during the Update portion, by collecting the results in LateUpdate.